### PR TITLE
Split (move) generic stuff for upgrades to the upgrade/common repo from el7toel8

### PR DESCRIPTION
**Pull request by pirat89**
This is PoC right now. With still a lot of work that needs to be done. See the `splitcomms` file for some notes we (@Rezney ) made. This is even not done yet. Some crucial topics needs to be resolved (e.g. what to do with IPUConfig and various constants in shared libraries).

Currently we want to have inside the common repo just content that is going to be really used in multiple upgrades. Additional content can be moved to the common repo later when needed. As well, we need to sync with several teams to check whether some content, that we keep inside the common repo, should be kept specific just for particular upgrade.

This is preparation step to be able to introduce the `el8toel9` leapp repository.

Sorry for making the msg less informative. I will polish it when switching from the draft to PR.

### CODE REVIEW

**Comment by pirat89** [comment]
Tested manually, everything seems to work as expected so far.

File: repos/system_upgrade/common/libraries/fetch.py [hunk 1]
```diff
@@ -29,12 +29,12 @@ def _request_data(service_path, cert, proxies, timeout=REQUEST_TIMEOUT):
         attempt += 1
         try:
             return requests.get(service_path, cert=cert, proxies=proxies, timeout=REQUEST_TIMEOUT)
-        except requests.exceptions.Timeout as e:
```
Comment (vinzenz) [inline repos/system_upgrade/common/libraries/fetch.py:32]: I don't think that you want to revert these things. That's by mistake I assume

File: splitcomms [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  - RHUI (checkrhui refactoring, ....)
+  - enablerhsmreposonrhel8
+  - firewalldfacts - ebtables out.... contact eric
+  - multipathconfread??
+  - reportsettargetrelease (msgs)
+  - setuptargetrepos
+  - ipuconfig
+  - desktop - keep in common? (kde does not exist); gnome-session on el8
+
+- SAPHANA scanner: @vincenz is it working/valid on el8?
```
Comment (vinzenz) [inline splitcomms:31]: From what I know it's generic enough to work. But we can give it a try and fix it in a follow up.

File: repos/system_upgrade/common/actors/enablerhsmtargetrepos/actor.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
+from leapp.actors import Actor
+from leapp.libraries.actor import enablerhsmtargetrepos
+from leapp.models import UsedTargetRepositories
+from leapp.tags import FirstBootPhaseTag, IPUWorkflowTag
+
+
+class EnableRHSMReposOnRHEL8(Actor):
```
Comment (vinzenz) [inline repos/system_upgrade/common/actors/enablerhsmtargetrepos/actor.py:7]:
```suggestion
class EnableRHSMTargetRepos(Actor):
```

File: repos/system_upgrade/common/actors/enablerhsmtargetrepos/actor.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    be enabled specifically using the subscription-manager (RHSM) utility. In case some custom repo was used
+    during the upgrade transaction, this won't be enabled by this actors as it is unknown to the subscription-manager.
+
+    We need to overwrite any RHSM release that may have been set before the upgrade, e.g. 7.6. Reasons:
+    - If we leave the old source RHEL release set, dnf calls on the upgraded target RHEL would fail.
+    - If we merely unset the release, users might end up updating the system to a newer version than the upgrade
+      was supposed to be upgrading to.
+    """
+
+    name = 'enable_rhsm_repos_on_rhel8'
```
Comment (vinzenz) [inline repos/system_upgrade/common/actors/enablerhsmtargetrepos/actor.py:23]:
```suggestion
    name = 'enable_rhsm_target_repos'
```

File: repos/system_upgrade/common/libraries/config/version.py [hunk 5]
```diff
@@ -11,8 +11,47 @@
...
+_SUPPORTED_VERSIONS = {
+    # Note: 'rhel-alt' is detected when on 'rhel' with kernel 4.x
+    '7': {'rhel': ['7.9'], 'rhel-alt': ['7.6'], 'rhel-saphana': ['7.7']},
+    '8': {'rhel': ['8.5', '8.6']},
+}
+
+
+def get_source_major_version():
+    return api.current_actor().configuration.version.source.split('.')[0]
+
```
Comment (vinzenz) [inline repos/system_upgrade/common/libraries/config/version.py:23]:
```suggestion

def get_target_major_version():
    return api.current_actor().configuration.version.target.split('.')[0]

```

File: repos/system_upgrade/common/libraries/dnfplugin.py [hunk 6]
```diff
@@ -8,8 +8,32 @@
 from leapp.libraries.common import guards, mounting, overlaygen, rhsm, utils
```
Comment (vinzenz) [inline repos/system_upgrade/common/libraries/dnfplugin.py:8]:
```suggestion
from leapp.libraries.common import guards, mounting, overlaygen, rhsm, utils
from leapp.libraries.common.config.version import get_target_major_version
```

File: repos/system_upgrade/common/libraries/dnfplugin.py [hunk 7]
```diff
@@ -8,8 +8,32 @@
...
 
+
 DNF_PLUGIN_NAME = 'rhel_upgrade.py'
-DNF_PLUGIN_PATH = os.path.join('/lib/python3.6/site-packages/dnf-plugins', DNF_PLUGIN_NAME)
+
+
+def get_target_major_version():
+    return api.current_actor().configuration.version.target.split('.')[0]
+
+
```
Comment (vinzenz) [inline repos/system_upgrade/common/libraries/dnfplugin.py:18]:
```suggestion
```

File: repos/system_upgrade/common/libraries/dnfplugin.py [hunk 8]
```diff
@@ -35,6 +59,9 @@ def build_plugin_data(target_repoids, debug, test, tasks, on_aws):
     """
     Generates a dictionary with the DNF plugin data.
     """
+
+    target_major_version = api.current_actor().configuration.version.target.split('.')[0]
+
```
Comment (vinzenz) [inline repos/system_upgrade/common/libraries/dnfplugin.py:64]:
```suggestion
```

File: repos/system_upgrade/common/libraries/dnfplugin.py [hunk 9]
```diff
@@ -50,7 +77,7 @@ def build_plugin_data(target_repoids, debug, test, tasks, on_aws):
             'disable_repos': True,
             'enable_repos': target_repoids,
             'gpgcheck': False,
-            'platform_id': 'platform:el8',
+            'platform_id': 'platform:el{}'.format(target_major_version),
```
Comment (vinzenz) [inline repos/system_upgrade/common/libraries/dnfplugin.py:80]:
```suggestion
            'platform_id': 'platform:el{}'.format(get_target_major_version()),
```

File: repos/system_upgrade/common/libraries/dnfplugin.py [hunk 10]
```diff
@@ -166,6 +193,9 @@ def install_initramdisk_requirements(packages, target_userspace_info, used_repos
     """
     Performs the installation of packages into the initram disk
     """
+
+    target_major_version = api.current_actor().configuration.version.target.split('.')[0]
+
```
Comment (vinzenz) [inline repos/system_upgrade/common/libraries/dnfplugin.py:198]:
```suggestion
```

File: repos/system_upgrade/common/libraries/dnfplugin.py [hunk 11]
```diff
@@ -175,7 +205,7 @@ def install_initramdisk_requirements(packages, target_userspace_info, used_repos
             'install',
             '-y',
             '--nogpgcheck',
-            '--setopt=module_platform_id=platform:el8',
+            '--setopt=module_platform_id=platform:el{}'.format(target_major_version),
```
Comment (vinzenz) [inline repos/system_upgrade/common/libraries/dnfplugin.py:208]:
```suggestion
            '--setopt=module_platform_id=platform:el{}'.format(get_target_major_version()),
```

File: repos/system_upgrade/common/libraries/fetch.py (hunk 1 above)
Comment (vinzenz) [reply repos/system_upgrade/common/libraries/fetch.py:32]: ++++ Still you need to change this back again.

**Review by vinzenz — CHANGES_REQUESTED** [review]
There are bunch of things still to be fixed.

**Comment by Rezney** [comment]
> There are bunch of things still to be fixed. Mainly there's still an accidental revert of repos/system_upgrade/el7toel8/libraries/fetch.py which probably should also live in `common` IMHO

Not sure If I am doing anything wrong but I see `fetch.py` still in the common (and it was there since the initial commits):

https://github.com/pirat89/leapp-repository/blob/split-common/repos/system_upgrade/common/libraries/fetch.py

Could you please elaborate more on the problem?

File: repos/system_upgrade/common/libraries/fetch.py (hunk 1 above)
Comment (pirat89) [reply repos/system_upgrade/common/libraries/fetch.py]: @Rezney

Comment (Rezney) [reply repos/system_upgrade/common/libraries/fetch.py]: fixed...

**Comment by AloisMahdal** [comment]
With

 * *leapp-0.12.1-100.202105311312Z.e12a38e.master.el7_9*
 * *leapp-repository-0.14.0-0.202106221054Z.9023c36.9023c365_split_common.PR668.el7_9*

we get this inhibitor:

    Risk Factor: high (inhibitor)
    Title: The installed OS version is not supported for the in-place upgrade to the target RHEL version
    Summary: The supported OS releases for the upgrade process:
     RHEL-ALT 7.6
    RHEL-SAPHANA 7.7
    RHEL 7.9
    Key: 1c7a98849a747ec9890f04bf4321de7280970715
    ----------------------------------------

I've seen it on all platforms I've checked.

For example, this system is RHEL-7.9, normally subscribed to prod CDN:

      [..]
      [..]
    morft_cm__switch(): poolids="<redacted>"
    morft_cm__switch(): cdn="prod"
    morft_cm__switch(): baseurl="https://cdn.redhat.com"
    morft_cm__switch(): preparing handler: enable
    morft_cm__switch(): running handler: enable
    Registering to: subscription.rhsm.stage.redhat.com:443/subscription
    The system has been registered with ID: <redacted>
    The registered system name is: <redacted>
    0 subscriptions removed at the server.
    Successfully attached a subscription for: High Availability for IBM System z (L3 only)
    Successfully attached a subscription for: Red Hat Enterprise Linux for IBM System z, Premium
    morft_cm__switch(): --- begin command: subscription-manager status ---
    morft_cm__switch():stdout: |+-------------------------------------------+
    morft_cm__switch():stdout: |   System Status Details
    morft_cm__switch():stdout: |+-------------------------------------------+
    morft_cm__switch():stdout: |Overall Status: Current
    morft_cm__switch():stdout: |
    morft_cm__switch():stdout: |System Purpose Status: Not Specified
    morft_cm__switch():stdout: |
    morft_cm__switch(): --- end command; es=0 ---
    morft_cm__switch(): external command: yum -y clean all
    [2021-06-23T13:13:23-0400]   assert.PASS: switch to: rhsm.cdn dspec=RHEL-7.9-updates-20210608.0 arch=s390x rlock=false variant=Server
      [..]
      [..]

with these enabled:

    rhel-7-for-system-z-optional-rpms/7Server/s390x                              enabled: 19,296
    rhel-7-for-system-z-rpms/7Server/s390x                                       enabled: 27,329

and yum-updated (4 packages).

File: repos/system_upgrade/common/actors/enablerhsmtargetrepos/libraries/enablerhsmtargetrepos.py [hunk 12]
```diff
@@ -17,13 +17,13 @@ def set_rhsm_release():
         rhsm.set_release(mounting.NotIsolatedActions(base_dir='/'), target_version)
     except CalledProcessError as err:
         api.current_logger().warning('Unable to set the {0} release through subscription-manager. When using dnf,'
-                                     ' content of the latest RHEL 8 minor version will be downloaded.\n{1}'
-                                     .format(target_version, str(err)))
+                                     ' content of the latest RHEL {1} minor version will be downloaded.\n{2}'
+                                     .format(target_version, target_version.split('.')[0], str(err)))
```
Comment (vinzenz) [inline repos/system_upgrade/common/actors/enablerhsmtargetrepos/libraries/enablerhsmtargetrepos.py:21]: as a follow up: This probably should you get_target_major_version instead

Comment (pirat89) [reply repos/system_upgrade/common/actors/enablerhsmtargetrepos/libraries/enablerhsmtargetrepos.py:21]: good catch. agree that we should do it in a followup

File: repos/system_upgrade/common/actors/ipuworkflowconfig/libraries/ipuworkflowconfig.py [hunk 13]
```diff
@@ -19,6 +17,30 @@
...
+    # unsupported fallback paths for RHEL 7
+    ('7', LEAPP_UPGRADE_FLAVOUR_DEFAULT): '8.4',
+    ('7', LEAPP_UPGRADE_FLAVOUR_SAP_HANA): '8.4',
+
+    # unsupported fallback paths for RHEL 8
+    ('8', LEAPP_UPGRADE_FLAVOUR_DEFAULT): '9.0',
+}
+
+
+def _get_major_version(version):
```
Comment (vinzenz) [inline repos/system_upgrade/common/actors/ipuworkflowconfig/libraries/ipuworkflowconfig.py:41]: Might be a useful helper function in libraries/config/version.py and can be used here
Ofc you can't use 'get_target_major_version' but this helper could be there, to keep the logic for version related functions in one place. 

Again, a follow up

File: repos/system_upgrade/common/actors/scaninstalledtargetkernelversion/libraries/scankernel.py [hunk 14]
```diff
@@ -9,10 +9,12 @@ def _get_kernel_version(kernel_name):
     except CalledProcessError:
         return ''
 
+    target_major_version = api.current_actor().configuration.version.target.split('.')[0]
+
```
Comment (vinzenz) [inline repos/system_upgrade/common/actors/scaninstalledtargetkernelversion/libraries/scankernel.py:13]:
```suggestion
```

File: repos/system_upgrade/common/actors/scaninstalledtargetkernelversion/libraries/scankernel.py [hunk 15]
```diff
@@ -9,10 +9,12 @@ def _get_kernel_version(kernel_name):
...
     for kernel in kernels:
         # name-version-release - we want the last two fields only
         version = '-'.join(kernel.split('-')[-2:])
-        if 'el8' in version:
+        if 'el{}'.format(target_major_version) in version:
```
Comment (vinzenz) [inline repos/system_upgrade/common/actors/scaninstalledtargetkernelversion/libraries/scankernel.py:17]:
```suggestion
        if 'el{}'.format(get_target_major_version()) in version:
```

Comment (vinzenz) [reply repos/system_upgrade/common/actors/scaninstalledtargetkernelversion/libraries/scankernel.py:17]: Needs ofc 
```python
from leapp.libraries.common.config.version import  get_target_major_version
```

File: repos/system_upgrade/common/actors/targetuserspacecreator/libraries/userspacegen.py [hunk 16]
```diff
@@ -125,19 +125,22 @@ def prepare_target_userspace(context, userspace_dir, enabled_repos, packages):
     """
     Implement the creation of the target userspace.
     """
+    target_major_version = api.current_actor().configuration.version.target.split('.')[0]
```
Comment (vinzenz) [inline repos/system_upgrade/common/actors/targetuserspacecreator/libraries/userspacegen.py:128]: again import get_target_major_version from leapp.libraries.config.version

File: repos/system_upgrade/common/actors/targetuserspacecreator/libraries/userspacegen.py [hunk 17]
```diff
@@ -489,15 +493,20 @@ def _copy_files(context, files):
             context.copy_to(file_task.src, file_task.dst)
 
 
+def _get_target_userspace():
+    target_major_version = api.current_actor().configuration.version.target.split('.')[0]
```
Comment (vinzenz) [inline repos/system_upgrade/common/actors/targetuserspacecreator/libraries/userspacegen.py:497]: again get_target_major_version from leapp.libraries.config.version

**Review by vinzenz — APPROVED** [review]
I am ok with the merge - Follow up tasks are required

**Comment by pirat89** [comment]
@vinzenz the task to fix the mentioned problems has been created: OAMG-5162

**Comment by pirat89** [comment]
We have to already merge it. Regarding the current results, test results seems to be ok, so let's believe the rest will be ok as well regarding the changes in this PR. Merging.

**Pull request merged**